File: app/src/main/java/io/github/anonymous/pepper_realtime/HumanDetectionAdapter.java
package io.github.anonymous.pepper_realtime;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.List;

/**
 * Adapter for displaying detected humans in the perception dashboard
 */
public class HumanDetectionAdapter extends RecyclerView.Adapter<HumanDetectionAdapter.HumanViewHolder> {
    
    private final List<PerceptionData.HumanInfo> humans = new ArrayList<>();
    
    public void updateHumans(List<PerceptionData.HumanInfo> newHumans) {
        // Use granular notifications based on size delta
        int oldSize = this.humans.size();
        int newSize = (newHumans == null) ? 0 : newHumans.size();

        if (newHumans == null || newHumans.isEmpty()) {
            // Remove all items if any
            if (oldSize > 0) {
                this.humans.clear();
                notifyItemRangeRemoved(0, oldSize);
            }
            return;
        }

        // Replace backing list
        this.humans.clear();
        this.humans.addAll(newHumans);

        if (oldSize == 0) {
            notifyItemRangeInserted(0, newSize);
        } else if (newSize == oldSize) {
            notifyItemRangeChanged(0, newSize);
        } else if (newSize > oldSize) {
            notifyItemRangeChanged(0, oldSize);
            notifyItemRangeInserted(oldSize, newSize - oldSize);
        } else { // newSize < oldSize
            notifyItemRangeChanged(0, newSize);
            notifyItemRangeRemoved(newSize, oldSize - newSize);
        }
    }
    
    @NonNull
    @Override
    public HumanViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_human_detection, parent, false);
        return new HumanViewHolder(view);
    }
    
    @Override
    public void onBindViewHolder(@NonNull HumanViewHolder holder, int position) {
        PerceptionData.HumanInfo human = humans.get(position);
        holder.bind(human);
    }
    
    @Override
    public int getItemCount() {
        return humans.size();
    }
    
    static class HumanViewHolder extends RecyclerView.ViewHolder {
        private final ImageView facePicture;
        private final TextView demographics;
        private final TextView distance;
        private final TextView emotion;
        private final TextView pleasure;
        private final TextView excitement;
        private final TextView smile;
        private final TextView attention;
        private final TextView engagement;
        // Azure UI Elements
        private final TextView azureHeadPose;
        private final TextView azureGlasses;
        private final TextView azureMask;
        private final TextView azureQuality;
        private int lastW = -1, lastH = -1;
        
        public HumanViewHolder(@NonNull View itemView) {
            super(itemView);
            facePicture = itemView.findViewById(R.id.human_face_picture);
            demographics = itemView.findViewById(R.id.human_demographics);
            distance = itemView.findViewById(R.id.human_distance);
            emotion = itemView.findViewById(R.id.human_emotion);
            pleasure = itemView.findViewById(R.id.human_pleasure);
            excitement = itemView.findViewById(R.id.human_excitement);
            smile = itemView.findViewById(R.id.human_smile);
            attention = itemView.findViewById(R.id.human_attention);
            engagement = itemView.findViewById(R.id.human_engagement);
            // Azure Views
            azureHeadPose = itemView.findViewById(R.id.azure_head_pose);
            azureGlasses = itemView.findViewById(R.id.azure_glasses);
            azureMask = itemView.findViewById(R.id.azure_mask);
            azureQuality = itemView.findViewById(R.id.azure_quality);
        }
        
        public void bind(PerceptionData.HumanInfo human) {
            // Set face picture with debugging
            if (human.facePicture != null) {
                facePicture.setImageBitmap(human.facePicture);
                int w = human.facePicture.getWidth();
                int h = human.facePicture.getHeight();
                if (w != lastW || h != lastH) {
                    android.util.Log.d("HumanAdapter", "🖼️ Displaying face picture for human " + human.id +
                        " (" + w + "x" + h + ")");
                    lastW = w; lastH = h;
                }
            } else {
                // Use default placeholder 
                facePicture.setImageResource(android.R.drawable.ic_menu_gallery);
                android.util.Log.d("HumanAdapter", "🖼️ Using placeholder for human " + human.id + " (no face picture available)");
            }
            
            // Set demographics
            demographics.setText(human.getDemographics());
            
            // Set distance
            distance.setText(human.getDistanceString());
            
            // Set basic emotion
            emotion.setText(human.getBasicEmotionDisplay());
            
            // Set pleasure state
            pleasure.setText(human.getPleasureStateDisplay());
            
            // Set excitement state
            excitement.setText(human.getExcitementStateDisplay());
            
            // Set smile state
            smile.setText(human.getSmileStateDisplay());
            
            // Set attention state
            attention.setText(human.getAttentionLevel());
            
            // Set engagement state
            engagement.setText(human.getEngagementLevel());

            // --- Bind Azure Data ---
            azureHeadPose.setText(formatHeadPose(human.azureYawDeg));
            azureGlasses.setText(String.format("Glasses: %s", human.glassesType));
            azureMask.setText(formatMask(human.isMasked));
            azureQuality.setText(String.format("Quality: %s", human.imageQuality));
        }

        private String formatHeadPose(Double yawDeg) {
            if (yawDeg == null) {
                return "Pose: N/A";
            }
            String direction;
            if (yawDeg < -10.0) {
                direction = "Right";
            } else if (yawDeg > 10.0) {
                direction = "Left";
            } else {
                direction = "Forward";
            }
            return String.format(java.util.Locale.US, "Pose: %s (%.0f°)", direction, yawDeg);
        }

        private String formatMask(Boolean isMasked) {
            if (isMasked == null) {
                return "Mask: N/A";
            }
            return "Mask: " + (isMasked ? "Yes" : "No");
        }
    }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/ui/HumanDetectionAdapter.kt
package io.github.anonymous.pepper_realtime.ui

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import io.github.anonymous.pepper_realtime.R
import io.github.anonymous.pepper_realtime.data.PerceptionData
import java.util.Locale

/**
 * Adapter for displaying detected humans in the perception dashboard
 */
class HumanDetectionAdapter : RecyclerView.Adapter<HumanDetectionAdapter.HumanViewHolder>() {

    private val humans = mutableListOf<PerceptionData.HumanInfo>()

    fun updateHumans(newHumans: List<PerceptionData.HumanInfo>?) {
        // Use granular notifications based on size delta
        val oldSize = humans.size
        val newSize = newHumans?.size ?: 0

        if (newHumans.isNullOrEmpty()) {
            // Remove all items if any
            if (oldSize > 0) {
                humans.clear()
                notifyItemRangeRemoved(0, oldSize)
            }
            return
        }

        // Replace backing list
        humans.clear()
        humans.addAll(newHumans)

        when {
            oldSize == 0 -> notifyItemRangeInserted(0, newSize)
            newSize == oldSize -> notifyItemRangeChanged(0, newSize)
            newSize > oldSize -> {
                notifyItemRangeChanged(0, oldSize)
                notifyItemRangeInserted(oldSize, newSize - oldSize)
            }
            else -> { // newSize < oldSize
                notifyItemRangeChanged(0, newSize)
                notifyItemRangeRemoved(newSize, oldSize - newSize)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HumanViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_human_detection, parent, false)
        return HumanViewHolder(view)
    }

    override fun onBindViewHolder(holder: HumanViewHolder, position: Int) {
        holder.bind(humans[position])
    }

    override fun getItemCount(): Int = humans.size

    class HumanViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val facePicture: ImageView = itemView.findViewById(R.id.human_face_picture)
        private val demographics: TextView = itemView.findViewById(R.id.human_demographics)
        private val distance: TextView = itemView.findViewById(R.id.human_distance)
        private val emotion: TextView = itemView.findViewById(R.id.human_emotion)
        private val pleasure: TextView = itemView.findViewById(R.id.human_pleasure)
        private val excitement: TextView = itemView.findViewById(R.id.human_excitement)
        private val smile: TextView = itemView.findViewById(R.id.human_smile)
        private val attention: TextView = itemView.findViewById(R.id.human_attention)
        private val engagement: TextView = itemView.findViewById(R.id.human_engagement)

        // Azure UI Elements
        private val azureHeadPose: TextView = itemView.findViewById(R.id.azure_head_pose)
        private val azureGlasses: TextView = itemView.findViewById(R.id.azure_glasses)
        private val azureMask: TextView = itemView.findViewById(R.id.azure_mask)
        private val azureQuality: TextView = itemView.findViewById(R.id.azure_quality)

        private var lastW = -1
        private var lastH = -1

        fun bind(human: PerceptionData.HumanInfo) {
            // Set face picture with debugging
            val bitmap = human.facePicture
            if (bitmap != null) {
                facePicture.setImageBitmap(bitmap)
                val w = bitmap.width
                val h = bitmap.height
                if (w != lastW || h != lastH) {
                    Log.d(TAG, "🖼️ Displaying face picture for human ${human.id} (${w}x${h})")
                    lastW = w
                    lastH = h
                }
            } else {
                // Use default placeholder
                facePicture.setImageResource(android.R.drawable.ic_menu_gallery)
                Log.d(TAG, "🖼️ Using placeholder for human ${human.id} (no face picture available)")
            }

            // Set demographics
            demographics.text = human.getDemographics()

            // Set distance
            distance.text = human.getDistanceString()

            // Set basic emotion
            emotion.text = human.getBasicEmotionDisplay()

            // Set pleasure state
            pleasure.text = human.getPleasureStateDisplay()

            // Set excitement state
            excitement.text = human.getExcitementStateDisplay()

            // Set smile state
            smile.text = human.getSmileStateDisplay()

            // Set attention state
            attention.text = human.getAttentionLevel()

            // Set engagement state
            engagement.text = human.getEngagementLevel()

            // --- Bind Azure Data ---
            azureHeadPose.text = formatHeadPose(human.azureYawDeg)
            azureGlasses.text = "Glasses: ${human.glassesType}"
            azureMask.text = formatMask(human.isMasked)
            azureQuality.text = "Quality: ${human.imageQuality}"
        }

        private fun formatHeadPose(yawDeg: Double?): String {
            if (yawDeg == null) {
                return "Pose: N/A"
            }
            val direction = when {
                yawDeg < -10.0 -> "Right"
                yawDeg > 10.0 -> "Left"
                else -> "Forward"
            }
            return String.format(Locale.US, "Pose: %s (%.0f°)", direction, yawDeg)
        }

        private fun formatMask(isMasked: Boolean?): String {
            return when (isMasked) {
                null -> "Mask: N/A"
                true -> "Mask: Yes"
                false -> "Mask: No"
            }
        }

        companion object {
            private const val TAG = "HumanAdapter"
        }
    }
}

